### ISSUE TITLE
default language missing when querying "translations"

### ISSUE THREAD

**Issue by danielzzz** [AD7six/cakephp-shadow-translate#23]
hi, I am not sure if it's a bug or I am doing something wrong. 

I have two languages in my app es_ES and es_CA, where the first one is set as default.

I configure I18n::locale('es_ES') and then I can receive both languages using find('translations')
then $entity->field - contains the original (es_ES) content and 
$entity->_translations['es_CA']->field has the catalan translation (good!)

but when I change locale globally to 'es_CA' using I18n::locale('es_CA') then both
$entity->field and $entity->_translations['es_CA']->field have the same value (which is ok, because we expect catalan to be the primary content), but es_ES which is in the base table is not received at all.

I expected these values from the base table to appear in _translations['es_ES'] after they are being overwritten when getting the translated content

thanks, best, dan

**Comment by AD7six** [comment]
> I expected these values from the base table to appear in _translations['es_ES']

I'm not really clear what you're doing/expecting - showing your code and some sample data would help.

It sounds like you're using the shadow-translate table for your primary language content - which is _not_ how the behavior is designed to be used; you shouldn't be accessing `$entity->_translations['es_CA']` at all really.

You should have e.g.:

```
mysql> select * from posts;
+----+-----------+---------------------------+---------------------+----------+
| id | title     | body                      | created             | modified |
+----+-----------+---------------------------+---------------------+----------+
|  1 | The title | Volem català per defecte  | 2015-01-19 17:31:14 | NULL     |
+----+-----------+---------------------------+---------------------+----------+
1 row in set (0.00 sec)

mysql> select * from posts_translations;
+----+--------+-------+---------------------------------+
| id | locale | title | body                            |
+----+--------+-------+---------------------------------+
|  1 | es_ES  | NULL  | Queremos castellano si se piden |
+----+--------+-------+---------------------------------+
```

If  Catalan is the default language, and that data is the translations table - it probably means the default locale is not set to `es_CA` (which you set in [the bootstrap file](https://github.com/cakephp/app/blob/master/config/bootstrap.php#L100)) - which is going to have various side effects - like there being no/wrong text in the main db tables, and the default language content being written to the translations table too.

**Comment by danielzzz** [comment]
thanks for your answer. 
You got me a bit wrong. I have the exact opposite situation: my primary content is in spanish and the translations table contains catalan - es_CA records. 

mysql> select * from posts;
+----+-----------+---------------------------+---------------------+----------+
| id | title     | body                      | created             | modified |
+----+-----------+---------------------------+---------------------+----------+
|  1 | The title | Castellano rulez  | 2015-01-19 17:31:14 | NULL     |
+----+-----------+---------------------------+---------------------+----------+
1 row in set (0.00 sec)

mysql> select * from posts_translations;
+----+--------+-------+---------------------------------+
| id | locale | title | body                            |
+----+--------+-------+---------------------------------+
|  1 | es_CA  | NULL  | Noltres volem català.. |
+----+--------+-------+---------------------------------+

ok, so when I have
I18n::locale('es_ES');
$posts = $this->Posts->find('translations')->toArray();

I get both es_ES and es_CA in the results (the primary table with es_ES and _translations with es_CA).
And this exactly what I expected.

but then when I use:
I18n:locale('es_CA'); //when a public user chooses another language.
$posts = $this->Posts->find('translations')->toArray();

I get just es_CA in both primary record and _translations.

I was hoping to be able to change the default language globally to something else but still receive old es_ES records.

**Comment by AD7six** [comment]
> I get just es_CA in both primary record and _translations.

Ok, that's to be expected - because the way the behavior (and the core) behavior works - it updates the main entity to be whatever is the applicable language for the request (that's probably stating the obvious) and the find-translations call is just an addition to that and so inherits that behavior. If you are interested in the implementation it's [the row mapper method](https://github.com/AD7six/cakephp-shadow-translate/blob/master/src/Model/Behavior/ShadowTranslateBehavior.php#L310-L365) that is responsible, i.e. this bit:

    if ($translation[$field] !== null) {
        if ($allowEmpty || $translation[$field] !== '') {
            $row[$field] = $translation[$field];
        }
    }

I'll have a think about whether it's appropriate to add/provide the untranslated values.. At the moment this and the core behavior (afaik/afair) don't provide any means to see the untranslated text if there is a translation record found.

**Comment by AD7six** [comment]
> I was hoping to be able to change the default language globally to something else but still receive old es_ES records.

Just for my own understanding of your situation - what's the reason/context for wanting that?

**Comment by ADmad** [comment]
@AD7six  I have encountered such a situation where client wakes up one day and decides he wants to use another language as default language which makes things a bit problematic since the default language is stored differently than other languages. 

One way to mitigate this would be configure the default language for behavior to be "empty" so that all languages end in the shadow table, though I doubt it would be that simple and other changes might be required too.

**Comment by AD7six** [comment]
@ADmad yeah.. but that makes things a bit awkard to use - I'd rather fix scenarios like that with a one-time update along the lines of:

    insert into foos_translations select id, "lang", bar from foos;
    update foos set bar = (select bar from foos_translations where foos_translations = foos.id)

I can guess at the usefulness of having access to the untranslated content - though I haven't had the need myself.

**Comment by danielzzz** [comment]
no, the default lang stays the same. 
I have two areas - admin and public.
The first shows all languages (for edition) the second just the active one.
I manage the change of the active one using "I18n::locale('xx_YY');
this way on the public site I can use entities as if they had just one language ($entity->field)

within the admin panel I receive all languages and create fields for edition, so the admin can edit them all at once. it works fine but only if the default language is not changed. in that case I have the problem described above.
right now I solved it hard setting the locale to spanish in the control panel, but I wasn't sure if the situation I reported is how you envisioned it to work.

the bottom line is if it's possible to retrieve the spanish (default) content having the locale set to another one.

**Comment by AD7six** [comment]
@danielzzz ADmad's comments were not about your situation - it was a different scenario in support of putting default-language content in the translation table.

Is your admin interface localized? 

I think implicitly the core behavior, which of course this repo mimics, implicitly assumes that retrive-all-translations is always done in the default language - I'm still undecided as to whether that's really a problem.

**Comment by danielzzz** [comment]
@AD7six yep, so am I... for now I have found that workaround, so it's not a big deal to me atm. 99% of the cases the admin interface can be in just one, base language. I can live with that (or better said my customers can ;) )

**Comment by AD7six** [comment]
Thanks for your interest here, as I no longer use or maintain this code I'll be archiving the project :)
